### DISCUSSION TITLE
Blazor 生命周期

### DISCUSSION THREAD

**Discussion by YSMC-W** [dotnetcore/BootstrapBlazor#765] [Quick start] [answered]
官方文档：[传送门](https://learn.microsoft.com/zh-cn/aspnet/core/blazor/components/lifecycle?view=aspnetcore-7.0&WT.mc_id=DT-MVP-5005089)。

组件生命周期事件：

1、如果组件是第一次呈现在请求上：

- 创建组件的实例。

- 执行属性注入。 运行 [SetParametersAsync](https://learn.microsoft.com/zh-cn/aspnet/core/blazor/components/lifecycle?view=aspnetcore-7.0#when-parameters-are-set-setparametersasync&WT.mc_id=DT-MVP-5005089)。
- 调用 [OnInitialized{Async}](https://learn.microsoft.com/zh-cn/aspnet/core/blazor/components/lifecycle?view=aspnetcore-7.0#component-initialization-oninitializedasync&WT.mc_id=DT-MVP-5005089)。 如果返回不完整的 [Task](https://learn.microsoft.com/zh-cn/dotnet/api/system.threading.tasks.task?WT.mc_id=DT-MVP-5005089)，则将等待 [Task](https://learn.microsoft.com/zh-cn/dotnet/api/system.threading.tasks.task?WT.mc_id=DT-MVP-5005089)，然后重新呈现组件。

2、调用 [OnParametersSet{Async}](https://learn.microsoft.com/zh-cn/aspnet/core/blazor/components/lifecycle?view=aspnetcore-7.0#after-parameters-are-set-onparameterssetasync&WT.mc_id=DT-MVP-5005089)。 如果返回不完整的 [Task](https://learn.microsoft.com/zh-cn/dotnet/api/system.threading.tasks.task?WT.mc_id=DT-MVP-5005089)，则将等待 [Task](https://learn.microsoft.com/zh-cn/dotnet/api/system.threading.tasks.task?WT.mc_id=DT-MVP-5005089)，然后重新呈现组件。
3、呈现所有同步工作和完整的 [Task](https://learn.microsoft.com/zh-cn/dotnet/api/system.threading.tasks.task?WT.mc_id=DT-MVP-5005089)。
![image](https://user-images.githubusercontent.com/15371924/226191307-dd7262eb-6e73-46e3-a226-8aecb06d507d.png)

文档对象模型 (DOM) 事件处理：

1、运行事件处理程序。
2、如果返回不完整的 [Task](https://learn.microsoft.com/zh-cn/dotnet/api/system.threading.tasks.task?WT.mc_id=DT-MVP-5005089)，则将等待 [Task](https://learn.microsoft.com/zh-cn/dotnet/api/system.threading.tasks.task?WT.mc_id=DT-MVP-5005089)，然后重新呈现组件。
3、呈现所有同步工作和完整的 [Task](https://learn.microsoft.com/zh-cn/dotnet/api/system.threading.tasks.task?WT.mc_id=DT-MVP-5005089)。
![image](https://user-images.githubusercontent.com/15371924/226191388-6c05e07f-3676-4f74-bf2a-9195653e687e.png)

Render 生命周期：

1、避免对组件进行进一步的呈现操作：
- 在第一次呈现后。
- 当 [ShouldRender](https://learn.microsoft.com/zh-cn/aspnet/core/blazor/components/rendering?view=aspnetcore-7.0#suppress-ui-refreshing-shouldrender&WT.mc_id=DT-MVP-5005089) 为 false 时。

2、生成呈现树差异并呈现组件。
3、等待 DOM 更新。
4、调用 [OnAfterRender{Async}](https://learn.microsoft.com/zh-cn/aspnet/core/blazor/components/lifecycle?view=aspnetcore-7.0#after-component-render-onafterrenderasync&WT.mc_id=DT-MVP-5005089)。
![image](https://user-images.githubusercontent.com/15371924/226191399-7d951af9-c00f-450a-a936-3f7aaa97347d.png)

开发人员调用 [StateHasChanged](https://learn.microsoft.com/zh-cn/aspnet/core/blazor/components/lifecycle?view=aspnetcore-7.0#state-changes-statehaschanged&WT.mc_id=DT-MVP-5005089) 会产生呈现。 有关详细信息，请参阅 [ASP.NET Core Razor 组件呈现](https://learn.microsoft.com/zh-cn/aspnet/core/blazor/components/rendering?view=aspnetcore-7.0&WT.mc_id=DT-MVP-5005089)。

本文简化了组件生命周期事件处理的某些方面，以阐明复杂的框架逻辑。 可能需要访问 [ComponentBase 引用源](https://github.com/dotnet/aspnetcore/blob/main/src/Components/Components/src/ComponentBase.cs?WT.mc_id=DT-MVP-5005089)，以将自定义事件处理与 Blazor 的生命周期事件处理集成。 引用源中的代码注释包括有关未出现在本文或 [API 文档](https://learn.microsoft.com/zh-cn/dotnet/api/?WT.mc_id=DT-MVP-5005089)中的生命周期事件处理的其他注释。 请注意，Blazor 的生命周期事件处理随时间推移而变化，并且可能会在每次发布时发生变化，恕不另行通知。

**Comment by ArgoZhang** [comment, marked as answer]
### 相关视频

https://space.bilibili.com/660853738/search/video?keyword=%E7%94%9F%E5%91%BD%E5%91%A8%E6%9C%9F
